### PR TITLE
Fix userStoreCountService for new backend implementation

### PR DESCRIPTION
**Pull request by wijith7**
Pre-Requisites  - This PR can merge after the kernel is released with https://github.com/wso2/carbon-kernel/pull/2350

Edit: The dependent kernel fix is available from kernel 4.6.3 onwards(refer https://github.com/wso2/carbon-kernel/commit/516529ab8675c3c6e4aa436a136909965c0ced96)

### CODE REVIEW

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 1]
```diff
@@ -202,4 +173,58 @@ public Long countByClaimInDomain(String claimURI, String valueFilter, String dom
 
     }
 
+    private RealmConfiguration realmConfiguration = null;
```
Comment (madurangasiriwardena) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: get the variables to the top of the class.

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 2]
```diff
@@ -202,4 +173,58 @@ public Long countByClaimInDomain(String claimURI, String valueFilter, String dom
...
+     * @param claimURI claim uri
+     * @param valueFilter filter
+     * @return user count
+     * @throws UserStoreCounterException UserStoreCounterException
+     */
+    private long getUserCountWithClaims(String claimURI, String valueFilter) throws UserStoreCounterException {
+
+        try {
+            realmConfiguration = CarbonContext.getThreadLocalCarbonContext().getUserRealm().getRealmConfiguration();
+            return UserStoreCountDSComponent.getRealmService().getUserRealm(this.realmConfiguration).
```
Comment (madurangasiriwardena) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: If you check the places that retrieve the userstore manager in this repo, it is done line below.

`UserStoreCountDSComponent.getRealmService().getTenantUserRealm(tenantId).getUserStoreManager();`

Let's try to follow the same approach.

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 3]
```diff
@@ -202,4 +173,58 @@ public Long countByClaimInDomain(String claimURI, String valueFilter, String dom
...
+                    getUserStoreManager().getUserCountWithClaims(claimURI, valueFilter);
+
+        } catch (UserStoreException e) {
+            String ErrorMsg = "No user count is retrieved from the user store.";
```
Comment (madurangasiriwardena) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: How can you say the error is "No user count is retrieved from the user store."?
What you can say here is you get "this" error while doing "this" operation.

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 4]
```diff
@@ -202,4 +173,58 @@ public Long countByClaimInDomain(String claimURI, String valueFilter, String dom
...
+     * @param filter filter
+     * @return role count
+     * @throws UserStoreCounterException UserStoreCounterException
+     */
+    private long getCountRoles(String filter) throws UserStoreCounterException {
+
+        try {
+            realmConfiguration = CarbonContext.getThreadLocalCarbonContext().getUserRealm().getRealmConfiguration();
+            return UserStoreCountDSComponent.getRealmService().getUserRealm(this.realmConfiguration).
+                    getUserStoreManager().countRoles(filter);
```
Comment (madurangasiriwardena) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: comments added for above `getUserCountWithClaims` needs to be fixed for this method as well.

**Comment by tharindu-bandara** [comment]
@wijith7 

Can you verify whether a kernel released with the dependant fix is available and whether that kernel version is updated in the framework to support this PR?

**Comment by tharindu-bandara** [comment]
Dependent kernel fix is available with Edit: Fix is available with https://github.com/wso2/carbon-kernel/commit/516529ab8675c3c6e4aa436a136909965c0ced96

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 5]
```diff
@@ -71,32 +74,31 @@
      * @return the number of roles matching the filter by each domain
      */
     public PairDTO[] countRoles(String filter) throws UserStoreCounterException {
+
         Set<String> userStoreDomains = UserStoreCountUtils.getCountEnabledUserStores();
-        //add 3 more for the counts of Internal, Application domains
+        //add 2 more for the counts of Internal, Application domains
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Make the first letter of the comment capital, and end the comment with a full stop.

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 6]
```diff
@@ -71,32 +74,31 @@
...
             roleCounts[i] = new PairDTO(userStoreDomain, Long.toString(count));
             i++;
         }
-
-        roleCounts[i] =  new PairDTO(UserCoreConstants.INTERNAL_DOMAIN, String.valueOf(
-                UserStoreCountUtils.getInternalRoleCount(filter)));
-        roleCounts[++i] =  new PairDTO(InternalStoreCountConstants.APPLICATION_DOMAIN, String.valueOf(
-                UserStoreCountUtils.getApplicationRoleCount(filter)));
+        String internalDomainFilter = UserCoreConstants.INTERNAL_DOMAIN + "/" + filter;
+        String applicationDomainFilter = InternalStoreCountConstants.APPLICATION_DOMAIN + "/" + filter;
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Shall we user `UserCoreConstants.DOMAIN_SEPARATOR` instead hardcoded value?

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 7]
```diff
@@ -109,27 +111,24 @@
...
-                }
-            } else {
-                //no action
+            long count = -1L;
+            String filterWithDomain = getFilterWithDomain(userStoreDomain, valueFilter);
+
+            try {
+                count = getUserCountWithClaims(claimURI, filterWithDomain);
+            } catch (UserStoreCounterException e) {
+                log.error("Error while getting user count with claim : " + claimURI + " from user store domain : "
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Shall we add a comma after the parameter? Ex: `"Error while getting user count with claim : " + claimURI + ", from user`

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 8]
```diff
@@ -139,17 +138,10 @@
      * @param filter the filter for the user name. Use '*' to have all.
      * @return the number of users matching the filter only within this user store domain
      */
-    public Long countUsersInDomain(String filter, String domain) throws UserStoreCounterException {
+    public long countUsersInDomain(String filter, String domain) throws UserStoreCounterException {
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: What would be the idea behind the public API changes?

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 9]
```diff
@@ -181,27 +163,87 @@ public Long countRolesInDomain(String filter, String domain) throws UserStoreCou
...
+            throws UserStoreCounterException {
 
+        String filterWithDomain = getFilterWithDomain(domain, valueFilter);
+        return getUserCountWithClaims(claimURI, filterWithDomain);
     }
 
     /**
      * Get count enabled user stores.
+     *
      * @return
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Missing return description.

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 10]
```diff
@@ -181,27 +163,87 @@ public Long countRolesInDomain(String filter, String domain) throws UserStoreCou
...
      * @throws UserStoreCounterException
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Fullstop at the end.

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 11]
```diff
@@ -181,27 +163,87 @@ public Long countRolesInDomain(String filter, String domain) throws UserStoreCou
...
         return domains.toArray(new String[domains.size()]);
 
     }
 
+    /**
+     * Get User count.
+     *
+     * @param claimURI    claim uri
+     * @param valueFilter filter
+     * @return user count
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Missing return description, and Full stops at the end. Shall we add more details for the valueFilter parameter?

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 12]
```diff
@@ -181,27 +163,87 @@ public Long countRolesInDomain(String filter, String domain) throws UserStoreCou
...
+            realmConfiguration = CarbonContext.getThreadLocalCarbonContext().getUserRealm().getRealmConfiguration();
+            UserStoreManager userStoreManager =
+                    UserStoreCountDSComponent.getRealmService().getTenantUserRealm(tenantId).getUserStoreManager();
+            if (userStoreManager instanceof org.wso2.carbon.user.core.UserStoreManager) {
+                return ((org.wso2.carbon.user.core.UserStoreManager) userStoreManager).getUserCountWithClaims(claimURI,
+                        valueFilter);
+            } else {
+                throw new UserStoreCounterException("Error while retrieving User Store from core");
+            }
+
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Redundant new line?

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 13]
```diff
@@ -181,27 +163,87 @@ public Long countRolesInDomain(String filter, String domain) throws UserStoreCou
...
+        } catch (UserStoreException e) {
+            String errorMsg = "error while retrieving user count from user store";
+            throw new UserStoreCounterException(errorMsg, e);
+        }
+    }
+
+    /**
+     * Get role count.
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: refer https://github.com/wso2/carbon-identity-framework/pull/2585/files#r371080827

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 14]
```diff
@@ -181,27 +163,87 @@ public Long countRolesInDomain(String filter, String domain) throws UserStoreCou
...
+     *
+     * @param filter filter
+     * @return role count
+     * @throws UserStoreCounterException UserStoreCounterException
+     */
+    private long getCountRoles(String filter) throws UserStoreCounterException {
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Shall we change the method name for better readability? Ex: `getRoleCount`

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 15]
```diff
@@ -181,27 +163,87 @@ public Long countRolesInDomain(String filter, String domain) throws UserStoreCou
...
+            }
+
+        } catch (UserStoreException e) {
+            String errorMsg = "Error while retrieving role count from user store";
+            throw new UserStoreCounterException(errorMsg, e);
+        }
+    }
+
+    /**
+     * Get filter with domain
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: refer https://github.com/wso2/carbon-identity-framework/pull/2585/files#r371080827

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 16]
```diff
@@ -181,27 +163,87 @@ public Long countRolesInDomain(String filter, String domain) throws UserStoreCou
...
+     *
+     * @param domain Domain
+     * @param filter Filter
+     * @return domain/filter
+     */
+    private String getFilterWithDomain(String domain, String filter) {
+
+        return domain + "/" + filter;
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Use domain separator from user core constants.

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 17]
```diff
@@ -181,27 +163,87 @@ public Long countRolesInDomain(String filter, String domain) throws UserStoreCou
...
+        try {
+            realmConfiguration = CarbonContext.getThreadLocalCarbonContext().getUserRealm().getRealmConfiguration();
+            UserStoreManager userStoreManager =
+                    UserStoreCountDSComponent.getRealmService().getTenantUserRealm(tenantId).getUserStoreManager();
+            if (userStoreManager instanceof org.wso2.carbon.user.core.UserStoreManager) {
+                return ((org.wso2.carbon.user.core.UserStoreManager) userStoreManager).countRoles(filter);
+            } else {
+                throw new UserStoreCounterException("Error while retrieving role count from core");
+            }
+
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Redundant new line.

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 18]
```diff
@@ -181,27 +163,87 @@ public Long countRolesInDomain(String filter, String domain) throws UserStoreCou
...
+     * @param filter filter
+     * @return role count
+     * @throws UserStoreCounterException UserStoreCounterException
+     */
+    private long getCountRoles(String filter) throws UserStoreCounterException {
+
+        try {
+            realmConfiguration = CarbonContext.getThreadLocalCarbonContext().getUserRealm().getRealmConfiguration();
+            UserStoreManager userStoreManager =
+                    UserStoreCountDSComponent.getRealmService().getTenantUserRealm(tenantId).getUserStoreManager();
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Handle the NPE where getUserRealm() could be null

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 19]
```diff
@@ -181,27 +163,87 @@ public Long countRolesInDomain(String filter, String domain) throws UserStoreCou
...
+            if (userStoreManager instanceof org.wso2.carbon.user.core.UserStoreManager) {
+                return ((org.wso2.carbon.user.core.UserStoreManager) userStoreManager).countRoles(filter);
+            } else {
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Remove the redundant `else` check, and throw the exception by default. We need to add at least a debug log saying that the user store manager does not comply.

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 20]
```diff
@@ -181,27 +163,87 @@ public Long countRolesInDomain(String filter, String domain) throws UserStoreCou
...
+     * @throws UserStoreCounterException UserStoreCounterException
+     */
+    private long getCountRoles(String filter) throws UserStoreCounterException {
+
+        try {
+            realmConfiguration = CarbonContext.getThreadLocalCarbonContext().getUserRealm().getRealmConfiguration();
+            UserStoreManager userStoreManager =
+                    UserStoreCountDSComponent.getRealmService().getTenantUserRealm(tenantId).getUserStoreManager();
+            if (userStoreManager instanceof org.wso2.carbon.user.core.UserStoreManager) {
+                return ((org.wso2.carbon.user.core.UserStoreManager) userStoreManager).countRoles(filter);
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Does the `filter` is domain qualified? better to include that in the java doc comment.

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 21]
```diff
@@ -181,27 +163,87 @@ public Long countRolesInDomain(String filter, String domain) throws UserStoreCou
...
 
+    /**
+     * Get User count.
+     *
+     * @param claimURI    claim uri
+     * @param valueFilter filter
+     * @return user count
+     * @throws UserStoreCounterException UserStoreCounterException
+     */
+    private long getUserCountWithClaims(String claimURI, String valueFilter) throws UserStoreCounterException {
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: All the comments in  `private long getCountRoles(String filter)` applies here as well.

One suggestion to reduce the duplicate code: Use a single private method to get the valid `org.wso2.carbon.user.core.UserStoreManager`. Then call that within this method(or directly call the corresponding userstore manager method.)

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 22]
```diff
@@ -41,26 +48,22 @@
...
-                    log.error("Error while getting user count from user store domain : " + userStoreDomain, e);
-                }
-            } else {
-                //no action
+            String filterWithDomain = getFilterWithDomain(userStoreDomain, filter);
+            long count = -1L;
+            try {
+                count = getUserCountWithClaims(UserStoreCountUtils.USERNAME_CLAIM, filterWithDomain);
+            } catch (UserStoreCounterException e) {
+                log.error("Error while getting user count from user store domain : " + userStoreDomain, e);
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: The exception is ignored here(only an error log is present). Shouldn't we pass the exception without handling it? Any reason just to have a log?

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 23]
```diff
@@ -41,26 +46,23 @@
...
-                    log.error("Error while getting user count from user store domain : " + userStoreDomain, e);
-                }
-            } else {
-                //no action
+            String filterWithDomain = getFilterWithDomain(userStoreDomain, filter);
+            long count = -1L;
+            try {
+                count = getUserCountWithClaims(UserStoreCountUtils.USERNAME_CLAIM, filterWithDomain);
+            } catch (UserStoreCounterException e) {
+                String errorMsg = "Error while getting user count from user store domain : ";
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: `... store domain : ";`

Missing user store domain parameter.

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 24]
```diff
@@ -71,32 +73,32 @@
      * @return the number of roles matching the filter by each domain
      */
     public PairDTO[] countRoles(String filter) throws UserStoreCounterException {
+
         Set<String> userStoreDomains = UserStoreCountUtils.getCountEnabledUserStores();
-        //add 3 more for the counts of Internal, Application domains
+        //Add 2 more for the counts of Internal, Application domains.
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Add a space at the beginning of the comment(after the backslashes).

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 25]
```diff
@@ -71,32 +73,32 @@
...
-                try {
-                    count = counter.countRoles(filter);
-                } catch (UserStoreCounterException e) {
-                    log.error("Error while getting role count from user store domain : " + userStoreDomain, e);
-                }
-            } else {
-                //no action
+            long count = -1L;
+            String filterWithDomain = getFilterWithDomain(userStoreDomain, filter);
+
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Redundant new line.

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 26]
```diff
@@ -71,32 +73,32 @@
...
+            try {
+                count = getRoleCount(filterWithDomain);
+            } catch (UserStoreCounterException e) {
+                log.error("Error while getting role count from user store domain : " + userStoreDomain, e);
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: As the `UserStoreCounterException` is created within this component, we do not need to catch that here. Therefore we can omit the try-catch block.

File: components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java [hunk 27]
```diff
@@ -109,27 +111,24 @@
...
-                    count = counter.countClaim(claimURI, valueFilter);
-                } catch (UserStoreCounterException e) {
-                    log.error("Error while getting user count with claim : " + claimURI + " from user store domain : "
-                            + userStoreDomain, e);
-                }
-            } else {
-                //no action
+            long count = -1L;
+            String filterWithDomain = getFilterWithDomain(userStoreDomain, valueFilter);
+
```
Comment (tharindu-bandara) [inline components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java]: Redundant new line

**Pull request merged**